multi_partner_learning.py

```diff
@@ -0,0 +1,522 @@
+        self.nb_epochs_done = (es.stopped_epoch + 1) if es.stopped_epoch != 0 else self.epoch_count
+
+    def compute_test_score(self):
+        """Return the score on test data of a final aggregated model trained in a federated way on each partner"""
+
+        partners_list = self.partners_list
+        partners_count = self.partners_count
+
+        epoch_count = self.epoch_count
+        minibatch_count = self.minibatch_count
+        is_early_stopping = self.is_early_stopping
+
+        x_val, y_val = self.val_data
+        x_test, y_test = self.test_data
+
+        # First, if only one partner, fall back to dedicated single partner function
+        if partners_count == 1:
+            return self.compute_test_score_for_single_partner(partners_list[0])
+
+        # Else, continue onto a federated learning procedure
+        partners_list = sorted(partners_list, key=operator.attrgetter("id"))
+        logger.info(
+            f"## Training and evaluating model on partners with ids: {['#' + str(p.id) for p in partners_list]}")
+
+        # Initialize variables
+        model_to_evaluate, sequentially_trained_model = None, None
+        if self.learning_approach == 'seq':
+            sequentially_trained_model = utils.generate_new_cnn_model()
+
+        # Train model (iterate for each epoch and mini-batch)
+        for epoch_index in range(epoch_count):
+
+            self.epoch_index = epoch_index
+
+            # Clear Keras' old models (except if the approach is sequential and the model has to persist across epochs)
+            if self.learning_approach != 'seq':
+                clear_session()
+
+            # Split the train dataset in mini-batches
+            self.split_in_minibatches()
+
+            # Iterate over mini-batches and train
+            for minibatch_index in range(minibatch_count):
+
+                self.minibatch_index = minibatch_index
+
+                if self.learning_approach == 'fedavg':
+                    self.compute_collaborative_round_fedavg()
+
+                elif self.learning_approach == 'seq':
+                    self.compute_collaborative_round_sequential(sequentially_trained_model)
+
+                elif self.learning_approach == 'seqavg':
+                    self.compute_collaborative_round_seqavg()
+
+            # At the end of each epoch, evaluate the model for early stopping on a global validation set
+            # Even in the 'seq' approach we aggregate here...
+            # ... to minimize the dependence to the order of partners on last minibatch
+            self.prepare_aggregation_weights()
+            model_to_evaluate = self.build_model_from_weights(self.aggregate_model_weights())
+            model_evaluation = model_to_evaluate.evaluate(
+                x_val, y_val, batch_size=constants.DEFAULT_BATCH_SIZE, verbose=0,
+            )
+            self.score_matrix_collective_models[epoch_index, minibatch_count] = model_evaluation[1]
```

Comment:
collective model is in fact the aggregated model ? 
Maybe rename to `test_score_aggregated_model` ?
And for the loss: `loss_aggregated_model` ?

Reply:
Yes but no: it is for the aggregated models in `fedavg` and `seqavg` approaches. In the `seq` approach it is (almost always) the model passed sequentially from mini-batch to mini-batch and partner to partner.